src/Adapter/MSTest.CoreAdapter/Execution/TestExecutionManager.cs

```diff
@@ -313,26 +302,21 @@ private void ExecuteTestsInSource(IEnumerable<TestCase> tests, IRunContext runCo
         }
 
         [SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification = "Requirement is to handle errors in user specified run parameters")]
-        private void CacheSessionParameters(IRunContext runContext, ITestExecutionRecorder testExecutionRecorder)
+        private IDictionary<string, object> GetSessionParameters(IRunContext runContext, ITestExecutionRecorder testExecutionRecorder)
         {
             if (!string.IsNullOrEmpty(runContext?.RunSettings?.SettingsXml))
             {
                 try
                 {
-                    var testRunParameters = RunSettingsUtilities.GetTestRunParameters(runContext.RunSettings.SettingsXml);
-                    if (testRunParameters != null)
```

Comment:
To address the problem of key value already being present in the dictionary, we can clear the sessionParameters object here and keep rest of the flow as it.

Reply:
Clearing sessionParameters would happen every time in the event of test run parameters are used, the performance would be the same with the change I propose. The only optimization would be if test run parameters are not used and the collision can't occur.
Alternatively I could do this.sessionParameters[kvp.Key] = kvp.Value instead and rename the function back to CacheSessionParameters(), keeping the original flow. But as I mentioned above you will have bleed-over of properties between runs which could have indeterminately affect the run depending on the order of execution and how the tests are designed. But the performance shouldn't be impacted at all.

Reply:
@CodeCyclone : Across sessions it should be ok to re-calculate TestRunParameters. That could potentially even change when the runsettings is changed. But for one session we should just calculate session parameters once as opposed to doing so for every source. For 100 test assemblies in a session we would be recalculating the same data 100 times.
`this.sessionParameters[kvp.Key] = kvp.Value` as you stated wouldn't be the right thing to do here - the run carries forward data from the previous session that could lead to all sorts of problems. 
It would be great if you could clear the data from the previous session and reset it in this API as @jayaranigarg  suggested.